provider/plugin/plugin.go

```diff
@@ -0,0 +1,337 @@
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"sigs.k8s.io/external-dns/endpoint"
+	"sigs.k8s.io/external-dns/plan"
+
+	backoff "github.com/cenkalti/backoff/v4"
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	mediaTypeFormatAndVersion = "application/external.dns.plugin+json;version=1"
+	contentTypeHeader         = "Content-Type"
+	acceptHeader              = "Accept"
+	varyHeader                = "Vary"
+	maxRetries                = 5
+)
+
+var (
+	recordsErrorsGauge = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: "external_dns",
+			Subsystem: "plugin_provider",
+			Name:      "records_errors",
+			Help:      "Errors with Records method",
+		},
+	)
+	applyChangesErrorsGauge = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: "external_dns",
+			Subsystem: "plugin_provider",
+			Name:      "applychanges_errors",
+			Help:      "Errors with ApplyChanges method",
+		},
+	)
+	propertyValuesEqualErrorsGauge = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: "external_dns",
+			Subsystem: "plugin_provider",
+			Name:      "propertyvaluesequal_errors",
+			Help:      "Errors with PropertyValuesEqual method",
+		},
+	)
+	adjustEndpointsErrorsGauge = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: "external_dns",
+			Subsystem: "plugin_provider",
+			Name:      "adjustendpointsgauge_errors",
+			Help:      "Errors with AdjustEndpoints method",
+		},
+	)
+)
+
+type PluginProvider struct {
+	client          *http.Client
+	remoteServerURL *url.URL
+}
+
+type PropertyValuesEqualRequest struct {
+	Name     string `json:"name"`
+	Previous string `json:"previous"`
+	Current  string `json:"current"`
+}
+
+type PropertyValuesEqualResponse struct {
+	Equals bool `json:"equals"`
+}
+
+func init() {
+	prometheus.MustRegister(recordsErrorsGauge)
+	prometheus.MustRegister(applyChangesErrorsGauge)
+	prometheus.MustRegister(propertyValuesEqualErrorsGauge)
+	prometheus.MustRegister(adjustEndpointsErrorsGauge)
+}
+
+func NewPluginProvider(u string) (*PluginProvider, error) {
+	parsedURL, err := url.Parse(u)
+	if err != nil {
+		return nil, err
+	}
+
+	// negotiate API information
+	req, err := http.NewRequest("GET", u, nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set(acceptHeader, mediaTypeFormatAndVersion)
+
+	client := &http.Client{}
+	var resp *http.Response
+	err = backoff.Retry(func() error {
+		resp, err = client.Do(req)
+		if err != nil {
+			log.Printf("error connecting to plugin api: %v", err)
+		}
+		return err
+	}, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxRetries))
+
+	vary := resp.Header.Get(varyHeader)
+	contentType := resp.Header.Get(contentTypeHeader)
+
+	if vary != contentTypeHeader {
+		return nil, fmt.Errorf("wrong vary value returned from server: %s", vary)
+	}
+
+	if contentType != mediaTypeFormatAndVersion {
```

Comment:
Maybe split into "wrong object" and "wrong version", then we have it parsed here when we need it later.

Reply:
I think we can keep using strings, the new version can be `"application/external.dns.plugin+json;version=2"` and we can use that quite easily. I'll keep it like this for now.